File: src/types/Attribute.ts
import * as DB from "tsondb/schema/dsl"
import { NestedTranslationMap } from "./Locale.js"

export const Attribute = DB.Entity(import.meta.url, {
  name: "Attribute",
  namePlural: "Attributes",
  comment: "A base attribute of a character.",
  type: () =>
    DB.Object({
      position: DB.Required({
        comment: "The position of the attribute in lists. This has to be a unique value.",
        type: DB.Integer({ minimum: 0 }),
      }),
      translations: NestedTranslationMap(
        DB.Required,
        "Attribute",
        DB.Object({
          name: DB.Required({
            comment: "The attribute’s name.",
            type: DB.String({ minLength: 1 }),
          }),
          abbreviation: DB.Required({
            comment: "The abbreviation of the attribute’s name.",
            type: DB.String({ minLength: 1 }),
          }),
          description: DB.Required({
            comment: "The description of the attribute.",
            type: DB.String({ minLength: 1 }),
          }),
        }),
      ),
    }),
  instanceDisplayName: {},
  sortOrder: {
    keyPath: "position",
  },
  uniqueConstraints: [
    {
      keyPath: "position",
    },
    {
      entityMapKeyPath: "translations",
      keyPathInEntityMap: "name",
    },
    {
      entityMapKeyPath: "translations",
      keyPathInEntityMap: "abbreviation",
    },
  ],
})

File: Sources/OptolithDatabaseSchema/Types/Attribute.swift
import FileDB

/// A base attribute of a character.
@Model
public struct Attribute {
    /// All translations for the entry, identified by IETF language tag (BCP47).
    @Relationship(Locale.self)
    let translations: [Locale.ID: Translation]

    @Embedded
    struct Translation {  // AttributeTranslation
        /// The attribute’s name.
        @MinLength(1)
        let name: String

        /// The abbreviation of the attribute’s name.
        @MinLength(1)
        let abbreviation: String

        /// The description of the attribute.
        @MinLength(1)
        let description: String
    }
}
